fuel-vm/src/constraints.rs

```diff
@@ -0,0 +1,118 @@
+//! Types to help constrain inputs to functions to only what is used.
+use fuel_asm::PanicReason;
+use fuel_asm::Word;
+use fuel_types::ContractId;
+
+use crate::consts::VM_MAX_RAM;
+use crate::consts::VM_MEMORY_SIZE;
+use crate::prelude::RuntimeError;
+
+pub mod reg_key;
+
+#[derive(Clone)]
+/// A range of memory that has been checked that it fits into the VM memory.
+pub struct CheckedMemRange(core::ops::Range<usize>);
+
+#[derive(Clone)]
+/// A range of memory that has been checked that it fits into the VM memory.
+/// This range can be used to read a value of type `T` from memory.
+pub struct CheckedMemValue<T>(CheckedMemRange, core::marker::PhantomData<T>);
```

Comment:
It'd be nice if we had a `SerializedSize` trait in fuel-types we could use to constrain `T` with here. Then we could ensure that the correct length is always used.

Reply:
That trait can't be const though

Reply:
You can have associated constants on a trait though, made an example here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1519603c242a9a86820791fd669b1357

Reply:
I just made an issue for this because I have a trait that basically does this in my PR that removes unsafe and this should probably be handled in that pr or a follow up to it.